src/infra/ingest-datafusion/src/visitor.rs

```diff
@@ -0,0 +1,254 @@
+    head: Multihash,
+    maybe_source_name: Option<&'a str>,
+
+    next_block_flags: MetadataBlockTypeFlags,
+    data_slices: Vec<Multihash>,
+
+    maybe_schema: Option<SchemaRef>,
+    maybe_source_event: Option<MetadataEvent>,
+    maybe_set_vocab: Option<SetVocab>,
+
+    maybe_prev_checkpoint: Option<Multihash>,
+    maybe_prev_watermark: Option<DateTime<Utc>>,
+    maybe_prev_source_state: Option<SourceState>,
+    maybe_prev_offset: Option<u64>,
+}
+
+impl<'a> DataWriterDataFusionMetaDataStateVisitor<'a> {
+    pub fn new(head: Multihash, maybe_source_name: Option<&'a str>) -> Self {
+        Self {
+            head,
+            maybe_source_name,
+
+            next_block_flags: Flag::SET_DATA_SCHEMA
+                | Flag::ADD_DATA
+                | Flag::SET_POLLING_SOURCE
+                | Flag::DISABLE_POLLING_SOURCE
+                | Flag::ADD_PUSH_SOURCE
+                | Flag::DISABLE_PUSH_SOURCE
+                | Flag::SET_VOCAB
+                | Flag::SEED,
+            data_slices: Vec::new(),
+
+            maybe_schema: None,
+            maybe_source_event: None,
+            maybe_set_vocab: None,
+
+            maybe_prev_checkpoint: None,
+            maybe_prev_watermark: None,
+            maybe_prev_source_state: None,
+            maybe_prev_offset: None,
+        }
+    }
+
+    pub fn get_metadata_state(self) -> Result<DataWriterMetadataState, ScanMetadataError> {
+        let merge_strategy = match (&self.maybe_source_event, self.maybe_source_name) {
+            // Source found
+            (Some(e), _) => match e {
+                MetadataEvent::SetPollingSource(e) => Ok(e.merge.clone()),
+                MetadataEvent::AddPushSource(e) => Ok(e.merge.clone()),
+                _ => unreachable!(),
+            },
+            // No source defined - assuming append strategy
+            (None, None) => Ok(MergeStrategy::Append(MergeStrategyAppend {})),
+            // Source expected but not found
+            (None, Some(source)) => Err(SourceNotFoundError::new(
+                Some(source),
+                format!("Source '{source}' not found"),
+            )),
+        }?;
+
+        Ok(DataWriterMetadataState {
+            head: self.head,
+            schema: self.maybe_schema,
+            source_event: self.maybe_source_event,
+            merge_strategy,
+            vocab: self.maybe_set_vocab.unwrap_or_default().into(),
+            data_slices: self.data_slices,
+            prev_offset: self.maybe_prev_offset,
+            prev_checkpoint: self.maybe_prev_checkpoint,
+            prev_watermark: self.maybe_prev_watermark,
+            prev_source_state: self.maybe_prev_source_state,
+        })
+    }
+
+    fn handle_set_data_schema(&mut self, e: &SetDataSchema) -> Result<(), ScanMetadataError> {
+        if self.maybe_schema.is_none() {
+            self.maybe_schema = Some(e.schema_as_arrow().int_err()?);
+        }
+
+        Ok(())
+    }
```

Comment:
Since once we visit `SetDataSchema` you are un-setting the `Flag::SET_DATA_SCHEMA` - we should never see the older schemas again during iteration, so the `if self.maybe_schema.is_none()` check seems to be redundant.

Same applies to many other similar checks here.

Reply:
makes sense

updated methods:
- `handle_set_data_schema()`
- `handle_set_vocab()`
- `handle_set_polling_source()`

remain the same:
- `handle_add_push_source()`

it seems to be more complex logic, so I decided to leave it as is